app/src/main/java/io/homeassistant/companion/android/vehicle/DomainListScreen.kt

```diff
@@ -0,0 +1,94 @@
+package io.homeassistant.companion.android.vehicle
+
+import android.car.Car
+import android.car.drivingstate.CarUxRestrictionsManager
+import android.content.pm.PackageManager
+import android.os.Build
+import androidx.annotation.RequiresApi
+import androidx.car.app.CarContext
+import androidx.car.app.Screen
+import androidx.car.app.model.Action
+import androidx.car.app.model.GridTemplate
+import androidx.car.app.model.Template
+import androidx.lifecycle.lifecycleScope
+import io.homeassistant.companion.android.BuildConfig
+import io.homeassistant.companion.android.common.R
+import io.homeassistant.companion.android.common.data.integration.Entity
+import io.homeassistant.companion.android.common.data.integration.IntegrationRepository
+import io.homeassistant.companion.android.common.data.integration.domain
+import io.homeassistant.companion.android.common.data.prefs.PrefsRepository
+import io.homeassistant.companion.android.common.data.servers.ServerManager
+import io.homeassistant.companion.android.util.vehicle.getDomainList
+import io.homeassistant.companion.android.util.vehicle.nativeModeActionStrip
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.launch
+
+@RequiresApi(Build.VERSION_CODES.O)
+class DomainListScreen(
+    carContext: CarContext,
+    val serverManager: ServerManager,
+    val integrationRepository: IntegrationRepository,
+    private val serverId: StateFlow<Int>,
+    private val allEntities: Flow<Map<String, Entity<*>>>,
+    private val prefsRepository: PrefsRepository
+) : Screen(carContext) {
+
+    private val domains = mutableSetOf<String>()
+
+    init {
+        lifecycleScope.launch {
+            allEntities.collect { entities ->
+                val newDomains = entities.values
+                    .map { it.domain }
+                    .distinct()
+                    .filter { it in MainVehicleScreen.SUPPORTED_DOMAINS }
+                    .toSet()
+                if (newDomains.size != domains.size || newDomains != domains) {
+                    domains.clear()
+                    domains.addAll(newDomains)
+                    invalidate()
+                }
+            }
+        }
+    }
+    override fun onGetTemplate(): Template {
+        val isAutomotive = carContext.packageManager.hasSystemFeature(PackageManager.FEATURE_AUTOMOTIVE)
+        val screen = MainVehicleScreen(
+            carContext,
+            serverManager,
+            serverId,
+            allEntities,
+            prefsRepository
+        ) { }
+        val isDrivingOptimized = screen.car?.let {
```

Comment:
This doesn't work correctly because
 - `car` is `null` until MainVehicleScreen is resumed, which this new instance never is, so it uses `?: false` immediately
 - this doesn't invalidate when the driving optimized state changes while viewing the domains list

I think you have to copy the entire car UX restrictions setup with callback + is driving optimized code. Maybe make it an extension function in Screen or CarContext or a base class for all our screens?

Reply:
done :)